### PR TITLE
feat: add vulnerability trend

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/application/application_vunerability.go [hunk 1]
```diff
@@ -8,23 +8,30 @@ import (
...
+		if err := db.Insert(&dbVuln); err != nil {
+			return sdk.WrapError(err, "InsertVulnerability> Unable to insert vulnerabilities")
+		}
 	}
 	return nil
 }
 
-// LoadVulnerabilitiesByRun loads vulnerabilities for the given run
-func LoadVulnerabilitiesByRun(db gorp.SqlExecutor, nodeRunID int64) ([]sdk.Vulnerability, error) {
+// LoadLatestVulnerabilities load vulnerabilities for the given application
```
Comment (yesnault) [inline engine/api/application/application_vunerability.go]: doc is about LoadVulnerabilities

File: engine/api/workflow/dao_node_run.go [hunk 2]
```diff
@@ -94,11 +93,11 @@ func LoadNodeRun(db gorp.SqlExecutor, projectkey, workflowname string, number, i
 		r.Coverage = cov
 	}
 	if loadOpts.WithVulnerabilities {
-		vulns, errV := application.LoadVulnerabilitiesByRun(db, r.ID)
-		if errV != nil && errV != sdk.ErrNotFound {
-			return nil, sdk.WrapError(errV, "LoadNodeRun>Error loading vulnerabilities for run %d", r.ID)
+		vuln, errV := loadVulnerabilityReport(db, r.ID)
+		if err != nil {
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run.go]: if errV and not err

File: engine/api/workflow/dao_node_run_vulnerability.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+// HandleVulnerabilityReport calculate vulnerability trend and save report
+func HandleVulnerabilityReport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, nr *sdk.WorkflowNodeRun, workerReport sdk.VulnerabilityWorkerReport) error {
+	var defaultBranch string
+	// Get default branch
+	if nr.VCSServer != "" {
+		// Get vcs info to known if we are on the default branch or not
+		projectVCSServer := repositoriesmanager.GetProjectVCSServer(proj, nr.VCSServer)
+		client, erra := repositoriesmanager.AuthorizedClient(db, cache, projectVCSServer)
+		if erra != nil {
+			return sdk.WrapError(sdk.ErrNoReposManagerClientAuth, "HandleVulnerabilityReport> Cannot get repo client %s : %s", nr.VCSServer, erra)
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run_vulnerability.go]: %v for erra

File: engine/api/workflow/dao_node_run_vulnerability.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+	// Get summary from previous run
+	previousRunReport, err := loadPreviousRunVulnerabilityReport(db, nr)
+	if err != nil && err != sdk.ErrNotFound {
+		return sdk.WrapError(err, "HandleVulnerabilityReport> Unable to get previous vulnerability report")
+	}
+	nodeRunReport.Report.PreviousRunSummary = previousRunReport
+
+	// Get summary from default branch
+	log.Warning("%s %s", defaultBranch, nr.VCSBranch)
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run_vulnerability.go]: log.warning??

File: engine/api/workflow/dao_node_run_vulnerability.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+func loadLatestRunVulnerabilityReport(db gorp.SqlExecutor, nr *sdk.WorkflowNodeRun, branch string) (map[string]int64, error) {
+	var dbReport dbNodeRunVulenrabilitiesReport
+	query := `
+    SELECT * FROM workflow_node_run_vulnerability
+    WHERE application_id = $1 AND workflow_id = $2 AND branch = $3
+    ORDER BY workflow_number DESC, workflow_node_run_id DESC
+    LIMIT 1
+  `
+	log.Warning("%d %d %s", nr.ApplicationID, nr.WorkflowID, branch)
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run_vulnerability.go]: log.Warning??

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5322.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api


























































































































































































  * Test_getWorkflowNodeRunHandler ✘


















* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow














































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript





* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5325.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api


























































































































































































  * Test_getWorkflowNodeRunHandler ✘


















* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow














































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript





* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5331.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api
























































































































































































  * Test_getLatestWorkflowRunHandler ✘

* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow














































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript





* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

File: engine/api/workflow/dao_node_run_vulnerability.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+	if err := db.SelectOne(&dbReport, query, nr.ApplicationID, nr.WorkflowID, branch); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, sdk.ErrNotFound
+		}
+		return nil, sdk.WrapError(err, "loadPreviousRunVulnerabilityReport> Unable to load previous report")
+	}
+	log.Warning("%+v", dbReport.Report.Summary)
```
Comment (yesnault) [inline engine/api/workflow/dao_node_run_vulnerability.go]: log.Warning??

**Comment by ovh-cds** [comment]
CDS Report ut-engine#5333.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api



























































































































































































  * Test_getWorkflowNodeRunHandler ✘


















* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/poller




* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/scheduler












* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker










* github.com/ovh/cds/engine/api/workflow














































* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk















* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript





* github.com/ovh/cds/sdk/plugin/dummy/caller


* github.com/ovh/cds/sdk/vcs/git

**Pull request merged**